File: macos/notch_overlay.m
#import <Cocoa/Cocoa.h>
#import <WebKit/WebKit.h>

@interface FloatingNotchPanel : NSPanel
@end

@implementation FloatingNotchPanel
- (BOOL)canBecomeKeyWindow { return YES; }
- (BOOL)canBecomeMainWindow { return YES; }
@end

@interface AppDelegate : NSObject <NSApplicationDelegate>
@property (strong) FloatingNotchPanel *panel;
@property (strong) WKWebView *webView;
@property (assign) BOOL isVisible;
@end

@implementation AppDelegate

- (void)applicationDidFinishLaunching:(NSNotification *)aNotification {
    [NSApp setActivationPolicy:NSApplicationActivationPolicyAccessory];

    NSScreen *screen = [NSScreen mainScreen];
    if (!screen) {
        NSLog(@"Error: No main screen found.");
        exit(1);
    }

    NSRect screenFrame = screen.frame;
    CGFloat width = 260.0;
    CGFloat height = 72.0;
    CGFloat xPos = (screenFrame.size.width - width) / 2.0;
    CGFloat yPos = screenFrame.size.height - height + 4.0; // Anchored directly below hardware notch

    NSRect contentRect = NSMakeRect(xPos, yPos, width, height);

    self.panel = [[FloatingNotchPanel alloc] initWithContentRect:contentRect
                                                       styleMask:NSWindowStyleMaskBorderless | NSWindowStyleMaskNonactivatingPanel
                                                         backing:NSBackingStoreBuffered
                                                           defer:NO];

    [self.panel setLevel:NSFloatingWindowLevel];
    [self.panel setOpaque:NO];
    [self.panel setBackgroundColor:[NSColor clearColor]];
    [self.panel setHasShadow:NO];
    [self.panel setIgnoresMouseEvents:NO];
    [self.panel setCollectionBehavior:NSWindowCollectionBehaviorCanJoinAllSpaces | 
                                      NSWindowCollectionBehaviorStationary | 
                                      NSWindowCollectionBehaviorIgnoresCycle];

    WKWebViewConfiguration *config = [[WKWebViewConfiguration alloc] init];
    self.webView = [[WKWebView alloc] initWithFrame:NSMakeRect(0, 0, width, height) configuration:config];
    [self.webView setValue:@NO forKey:@"drawsBackground"];
    [self.webView setAutoresizingMask:NSViewWidthSizable | NSViewHeightSizable];

    NSString *currentPath = [[NSFileManager defaultManager] currentDirectoryPath];
    NSString *htmlPath = [currentPath stringByAppendingPathComponent:@"stacky_notch_bar_runtime.html"];
    NSURL *fileURL = [NSURL fileURLWithPath:htmlPath];

    if ([[NSFileManager defaultManager] fileExistsAtPath:htmlPath]) {
        [self.webView loadFileURL:fileURL allowingReadAccessToURL:[fileURL URLByDeletingLastPathComponent]];
    } else {
        NSLog(@"Warning: stacky_notch_bar_runtime.html not found at %@", htmlPath);
    }

    [self.panel setContentView:self.webView];
    [self.panel orderFrontRegardless];
    self.isVisible = YES;

    printf("[+] Stacky Native macOS Notch Overlay running at (x: %.1f, y: %.1f)\n", xPos, yPos);
    printf("[+] Interactive commands: 'toggle', 'hide', 'show', 'exit'\n");
    fflush(stdout);

    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
        char buffer[256];
        while (fgets(buffer, sizeof(buffer), stdin)) {
            NSString *input = [[NSString stringWithUTF8String:buffer] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]].lowercaseString;
            dispatch_async(dispatch_get_main_queue(), ^{
                if ([input isEqualToString:@"toggle"]) {
                    [self toggle];
                } else if ([input isEqualToString:@"hide"]) {
                    [self.panel orderOut:nil];
                    self.isVisible = NO;
                    printf("[+] Notch overlay hidden.\n");
                    fflush(stdout);
                } else if ([input isEqualToString:@"show"]) {
                    [self.panel orderFrontRegardless];
                    self.isVisible = YES;
                    printf("[+] Notch overlay shown.\n");
                    fflush(stdout);
                } else if ([input isEqualToString:@"exit"] || [input isEqualToString:@"quit"]) {
                    [NSApp terminate:nil];
                }
            });
        }
    });
}

- (void)toggle {
    if (self.isVisible) {
        [self.panel orderOut:nil];
        self.isVisible = NO;
        printf("[+] Notch overlay hidden.\n");
    } else {
        [self.panel orderFrontRegardless];
        self.isVisible = YES;
        printf("[+] Notch overlay shown.\n");
    }
    fflush(stdout);
}

@end

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSApplication *app = [NSApplication sharedApplication];
        AppDelegate *delegate = [[AppDelegate alloc] init];
        app.delegate = delegate;
        [app run];
    }
    return 0;
}

File: macos/notch_overlay.swift
import Cocoa
import WebKit

class FloatingNotchPanel: NSPanel {
    override var canBecomeKey: Bool { return true }
    override var canBecomeMain: Bool { return true }
}

class AppDelegate: NSObject, NSApplicationDelegate {
    var panel: FloatingNotchPanel!
    var webView: WKWebView!
    var isVisible: Bool = true

    func applicationDidFinishLaunching(_ notification: Notification) {
        NSApp.setActivationPolicy(.accessory) // Runs as a menu/accessory floating app without dock icon clutter

        guard let screen = NSScreen.main else {
            NSLog("Error: No main screen detected.")
            exit(1)
        }

        let screenWidth = screen.frame.width
        let screenHeight = screen.frame.height

        let panelWidth: CGFloat = 260
        let panelHeight: CGFloat = 72
        let xPos = (screenWidth - panelWidth) / 2
        let yPos = screenHeight - panelHeight + 4 // Clings directly beneath hardware notch

        let contentRect = NSRect(x: xPos, y: yPos, width: panelWidth, height: panelHeight)

        panel = FloatingNotchPanel(
            contentRect: contentRect,
            styleMask: [.borderless, .nonactivatingPanel],
            backing: .buffered,
            defer: false
        )

        panel.level = .floating
        panel.isOpaque = false
        panel.backgroundColor = .clear
        panel.hasShadow = false
        panel.ignoresMouseEvents = false
        panel.collectionBehavior = [.canJoinAllSpaces, .stationary, .ignoresCycle]

        // WebKit Configuration
        let config = WKWebViewConfiguration()
        config.preferences.setValue(true, forKey: "developerExtrasEnabled")
        
        webView = WKWebView(frame: NSRect(x: 0, y: 0, width: panelWidth, height: panelHeight), configuration: config)
        webView.setValue(false, forKey: "drawsBackground") // 100% transparency
        webView.autoresizingMask = [.width, .height]

        // Locate runtime HTML file
        let currentDir = FileManager.default.currentDirectoryPath
        let htmlPath = URL(fileURLWithPath: currentDir).appendingPathComponent("stacky_notch_bar_runtime.html")

        if FileManager.default.fileExists(atPath: htmlPath.path) {
            webView.loadFileURL(htmlPath, allowingReadAccessTo: htmlPath.deletingLastPathComponent())
        } else {
            NSLog("Warning: Could not find stacky_notch_bar_runtime.html at \(htmlPath.path)")
        }

        panel.contentView = webView
        panel.orderFrontRegardless()

        print("[+] Stacky Native Notch Overlay running at (x: \(xPos), y: \(yPos))")
        print("[+] Send 'toggle', 'hide', 'show', or 'exit' via stdin.")

        // Listen on background thread for stdin commands to toggle or control
        DispatchQueue.global(qos: .background).async { [weak self] in
            while let line = readLine() {
                let cmd = line.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
                DispatchQueue.main.async {
                    switch cmd {
                    case "toggle":
                        self?.togglePanel()
                    case "hide":
                        self?.panel.orderOut(nil)
                        self?.isVisible = false
                    case "show":
                        self?.panel.orderFrontRegardless()
                        self?.isVisible = true
                    case "exit", "quit":
                        NSApp.terminate(nil)
                    default:
                        break
                    }
                }
            }
        }
    }

    func togglePanel() {
        if isVisible {
            panel.orderOut(nil)
            isVisible = false
            print("[+] Notch overlay hidden.")
        } else {
            panel.orderFrontRegardless()
            isVisible = true
            print("[+] Notch overlay visible.")
        }
    }
}

let app = NSApplication.shared
let delegate = AppDelegate()
app.delegate = delegate
app.run()
